### PR TITLE
ms precision for settings and exponential back off

### PR DESCRIPTION
**Pull request by serioushaircut**
- [x] added setting for exponential back off with max delay value

### CODE REVIEW

**Comment by serioushaircut** [comment]
retest this please

**Comment by serioushaircut** [comment]
retest this please

**Comment by serioushaircut** [comment]
Those cucumber specs for redis failover are pretty flaky (at least now). I restarted the build with force-pushes several times only with changes to the examples but it worked only twice. Don't know whether a change of mine caused this.

**Comment by skaes** [comment]
The 1 second sleep was added to avoid busy looping for the case when a  message gets redelivered immediately. And this is exactly what's happening currently. When the client rejects a message, the broker will deliver it immediately either to the same client or some other client. If the specified delay for the message has not been reached yet, the client immediately rejects it again.

I'm not sure what the performance/resource utilization effect of reducing the sleep by a factor of 10 would have on our broker setup. Therefor I'm really reluctant to change it.

**Comment by skaes** [comment]
@serioushaircut the cucumber tests are flaky only on travis. We did not want to put more time into reducing the flakiness.  Running the tests in low traffic times usually leads to green builds.

File: lib/beetle/message.rb [hunk 1]
```diff
@@ -362,5 +364,18 @@ def ack!
...
+    def exponential_back_off?
+      @exp_bo
+    end
+
+    def calculate_next_delay(n, base_delay)
+      unless exponential_back_off?
+        base_delay
+      else
+        base_delay_n = base_delay * (2**n)
+        rand((base_delay_n - base_delay_n / 4.0)..(base_delay_n + base_delay_n / 2.0))
```
Comment (skaes) [inline lib/beetle/message.rb]: I'm not sure that picking a random number within the backoff interval has any effect and I'd rather keep the code simple. What's missing is an upper bound for the delay. A max_delay option. See ZMQ_RECONNECT_IVL and ZMQ_RECONNECT_IVL_MAX for zeromq http://api.zeromq.org/2-1:zmq-setsockopt.

**Comment by serioushaircut** [comment]
@skaes Thx for the quick review. Made the changes. But actually w/o below seconds resolution (I figured there was a reason for it being 1s and asked sebastian roebke about its consequences before I added it) and especially w/o jittering the backoff, I'd actually recommend not to introduce it at all :(. I don't think it would help anyone, at least I would rather not use it as it is now. 
I hope I didn't steal too much of your time.

**Comment by serioushaircut** [comment]
and the last travis build is half green, btw.

**Comment by skaes** [comment]
@serioushaircut I think exponential backoff is a useful thing to have, as stated already. I don't understand why you think that it will not be effective unless you also introduce jitter.

Maybe we should discuss what you're trying to achieve over a coffee.

**Comment by boosty** [comment]
> I'm not sure what the performance/resource utilization effect of reducing the sleep by a factor of 10 would have on our broker setup. Therefore I'm really reluctant to change it.

We could keep it to 1 sec by default, but make it configurable so that teams can use a lower value if it makes sense for them?

**Comment by serioushaircut** [comment]
@skaes should I do anything else? As it is now, it offers exp bo with one second resolution and a max bo can be set. The attempts files are mostly copy paste, should I do anything about them?

**Comment by skaes** [comment]
Thx a lot for the contribution. I will merge it. And we should talk about the ms resolution separately

**Comment by skaes** [comment]
released as 2.1.0 (with modifications).

**Pull request merged**